notes/quic.md

```diff
@@ -16,19 +16,88 @@ different streams/requests on the same connection, given one TCP connection
 
-QUIC was borne as a way to be more efficient on this model where multiple
+QUIC was born as a way to be more efficient on this model where multiple
 requests are made using the same underlying connection, which is reflected
 on its design. Here I document some core design decisions that caught my
 attention, not much because they are amazing by itself, but they present
 considerable differences to TCP and may be a better fit in some scenarios
-because of it.
+because of it (the main one being "better HTTP2").
+
 
 # Built on top of UDP
 
-# Always Confidential
+This one may be accidental, since the only protocols that will work on Internet
+infrastructure are TCP and UDP if you want to have a transport protocol that
+works on the "wild" you will need to do something on top of one of these two, or
+buy a huge fight like IPv6, which so far shows how hard it is to introduce
+new protocols at Internet scale. Since networks were designed for encapsulation
+it is fairly easy to build new protocols on top of UDP/TCP. Since TCP has
+guarantees that goes against the QUIC design, you are left with UDP as an option.
+
+There are two properties that emerge from this decision that I can think so far.
+One is that you will always have the option of using your own user space
+implementation of QUIC, even when kernels already implement it on the future, which
+seems like a nice/useful option that is much harder to do with UDP/TCP.
+
+Another one is that at least for user space implementations you don't have
+limitations on how much connection you can establish per process, beyond
+the CPU/Memory, since you are going to have just one file descriptor open
+for the UDP socket and all connections/streams are built on top of that
+on user space. I have no idea how that scales, but I'm kinda curious if it would not
+be feasible in that model for a process alone manage 100K connections, something
+that is not doable in TCP given the max number of file descriptors limitation.
+
 
 # Connection identity decoupled from IP
 
+As the time I'm writing this I still don't know much details on how
+this work, I just know that QUIC connections are decoupled from IP addresses
+in the sense that the identity of a connection is not tied to IP's, instead
+of the traditional quadruple **<ipx>:<portx>:<ipy>:<porty>**.
+
+I would guess it is related to TLS and that the enforcement of it on
+the protocol guarantees a way to know that after you start a connection
+any communication that happens on it is still being done between the same
+endpoints, even if its IP address change. So you can decouple from IP
+as a mean of identification but without giving up security. I'm not very
+good with security stuff but AFAIK it is possible to impersonate an IP
+depending on how a network is designed, so I'm not sure if using IPs as
+identity is a good idea anyway.
+
+The main property that came to my mind with this is that connections will be
```

Comment:
I wonder how this would work, though -- if the client's IP address changes, how does the server know which IP address to send the response packets to?

Reply:
Perhaps there is something on the protocol like if the IP address change the client sends an update to the server, like refreshing the connection (which is what you do by hand on TCP when the connection drops). But if both change IPs at the same time (like in a peer to peer scenario) in that case the connection will probably be dropped. Will take a read on the [RFC](https://tools.ietf.org/html/draft-ietf-quic-transport-27#section-5) to see if I can make sense of these scenarios. Perhaps I misunderstood tthe part that it is decoupled from IP addresses (actually I just saw a dude on a presentation say that =P)

Reply:
First update, from the [Connections section](https://tools.ietf.org/html/draft-ietf-quic-transport-27#section-5):

```
Once established,
a connection may migrate to a different IP or port at either endpoint
as described in Section 9.  Finally, a connection may be terminated
by either endpoint, as described in Section 10.
```

So I did got the behavior right, will dig for details

Reply:
From [connection migration](https://tools.ietf.org/html/draft-ietf-quic-transport-27#section-9) it seems that one of the endpoints can expect for a new valid path on the network to be created, so in a client server scenario I imagine that the server would start getting errors when trying to send data to the client and it can wait for the client to send something (not sure yet what, I imagine something to inform the change in IP/port) and then it can validate the new path and start using it.

Reply:
And from [path validation](https://tools.ietf.org/html/draft-ietf-quic-transport-27#section-8.2) it seems something like I was imagining, the peer that is migrating (changed IP/port) initiates a path validation:

```
Path validation is used during connection migration (see Section 9
   and Section 9.6) by the migrating endpoint to verify reachability of
   a peer from a new local address.
```

It feels like a new handshake, but I suppose it is more lightweight than the initial connection establishment because you don't need to reestablish cryptographic/TLS stuff.

Reply:
And it is literally some random stuff just to say "hey I have a new address":

```
To initiate path validation, an endpoint sends a PATH_CHALLENGE frame
   containing a random payload on the path to be validated.

   An endpoint MAY send multiple PATH_CHALLENGE frames to guard against
   packet loss.
```

Reply:
Ah, so they just designed a system to update the IP address. Pretty cool :)

Reply:
Yeah, it seems they model this as the **path**, so it is basically validating a new path (and nothing related to the path is used as a means for identity). Haven't found on the RFC info regarding simultaneous migration, I imagine that in that scenario it will just drop the connection since it seems that there is nothing to do =P (at least in a client/server scenario it is very unlikely)